src/dune_engine/scheduler.ml

```diff
@@ -494,6 +522,54 @@ end = struct
     pid
 end
 
+module Jobserver : sig
+  type t
+
+  val create : Event.Queue.t -> Env.t -> t option
+
+  val with_job : t -> f:(unit -> 'a Fiber.t) -> 'a Fiber.t
+end = struct
+  type t =
+    { jobserver : Jobserver.t
+    ; read : Task_queue.t
+    ; write : Task_queue.t
+    }
+
+  let create q env =
+    let open Option.O in
+    let scheduler =
+      let create_thread_safe_ivar () =
+        let ivar = Fiber.Ivar.create () in
+        let fill v =
+          Event.Queue.send_jobserver_completed q (Fiber.Fill (ivar, v))
+        in
+        Event.Queue.register_jobserver_started q;
+        (ivar, fill)
+      in
+      { Task_queue.Scheduler.create_thread_safe_ivar
+      ; spawn_thread = Thread.spawn
+      }
+    in
+    let+ jobserver = Jobserver.of_env env in
+    let read = Task_queue.create scheduler in
+    let write = Task_queue.create scheduler in
+    { jobserver; read; write }
+
+  let with_job { read; write; jobserver } ~f =
+    let open Fiber.O in
+    let* job =
+      Task_queue.task_exn read ~f:(fun () -> Jobserver.wait_next jobserver)
+    in
+    Fiber.finalize f ~finally:(fun () ->
+        let+ res =
+          Task_queue.task write ~f:(fun () -> Jobserver.Job.release job)
```

Comment:
We don't need to wait for the token to be written here. We should just queue it, signal the writer thread and move on.